feat: Add Qdrant module

russcam opened the pull request:
Closes #992

## What does this PR do?

This commit adds a qdrant container to the list of supported Testcontainers, for use with the [Qdrant vector database](https://qdrant.tech/).

The qdrant container allows configuration of:

- an API key to authenticate to Qdrant
- a certificate and private key in PEM format used to secure communication to Qdrant with Transport Layer Security

## Why is it important?

Provides a qdrant test container for folks that need it

## Related issues

#992
Supersedes #994

## How to test this PR

Run tests in Testcontainers.Qdrant.Tests

russcam committed:
Add Qdrant container

This commit adds a qdrant container to the list of supported
Testcontainers.

The qdrant container allows configuration of:

- an API key to authenticate to Qdrant
- an x509 certificate used to secure communication to Qdrant with
  Transport Layer Security
- a custom configuration file. See
  https://qdrant.tech/documentation/guides/configuration/

Closes #992

russcam committed:
Fix TLS tests on Linux

Qdrant uses rustls for TLS, which does not accept IPv4 or IPv6 addresses as the hostname in SNI, adhering
to RFC 6066: https://www.rfc-editor.org/rfc/rfc6066#page-7

Sending an IPv4 or IPv6 address as the hostname in SNI causes the handshake to be rejected and
qdrant to log

WARN rustls::msgs::handshake: Illegal SNI hostname received "127.0.0.1"

By default, .NET on linux sends the DNS name in the URI in Server Name Indication (SNI),
irrespective of whether it's an IP address or not. In order to resolve this, a custom
Host request header is added, per

https://github.com/dotnet/runtime/issues/20876#issuecomment-379585707

russcam committed:
Remove superfluous solution items

russcam committed:
PR feedback

russcam committed:
remove superfluous whitespace

russcam committed:
address PR checks

eerhardt commented on src/Testcontainers.Qdrant/Testcontainers.Qdrant.csproj:
```diff
@@ -0,0 +1,12 @@
+<Project Sdk="Microsoft.NET.Sdk">
+    <PropertyGroup>
+        <TargetFrameworks>net6.0;net8.0;netstandard2.0;netstandard2.1</TargetFrameworks>
```
Are all these really necessary? I would think the following would be enough:

```suggestion
        <TargetFrameworks>net6.0;netstandard2.0</TargetFrameworks>
```

russcam replied:
Every other TestContainer module defines these target frameworks, so was sticking to the project convention. I've also added net462 for same reason Milvus module does: https://github.com/testcontainers/testcontainers-dotnet/pull/1131#discussion_r1507712752

HofmeisterAn replied:
I think we should stick to the repo convention too 👍. In the end, it does not really matter, right (except for the legacy framework)? We can consider dropping support for all projects. I just kept it because there were some preprocessor directives we used in the past, but I think we have already changed them. I need to double-check.

eddumelendez commented on src/Testcontainers.Qdrant/QdrantBuilder.cs:
```diff
@@ -0,0 +1,77 @@
+			.WithPortBinding(QdrantHttpPort, true)
+			.WithPortBinding(QdrantGrpcPort, true)
+			.WithWaitStrategy(Wait.ForUnixContainer()
+				.UntilMessageIsLogged(".*Actix runtime found; starting in Actix runtime.*"));
```
shouldn't it wait for `/readyz` in port `6333`?

russcam replied:
I've changed the wait for strategy to use the readyz endpoint instead, handling the case where a certificate is provided. Ability to provide a certificate has to be supported for net462, because of the underlying gRPC packages used in net462- see [qdrant .NET SDK README](https://github.com/qdrant/qdrant-dotnet).

russcam committed:
Use readyz endpoint

This commit moves the wait for strategy to
use the readyz endpoint. If a certificate has
been specified, perform the check with TLS and
allow any certificate to pass validation.

russcam committed:
Also target net462

The qdrant gRPC client uses .NET Framework build
of Grpc.Net.Client, which uses WinHttpHandler. The
wiring up is easier on net462 if net462 is targeted
specifically. Same reason why the Milvus container
also targets net462 (https://github.com/testcontainers/testcontainers-dotnet/pull/1131#discussion_r1507712752).

HofmeisterAn commented on src/Testcontainers.Qdrant/QdrantConfiguration.cs:
```diff
@@ -0,0 +1,73 @@
+	/// </summary>
+	public QdrantConfiguration(string apiKey = null, string certificate = null, string privateKey = null)
+	{
+		ApiKey = apiKey;
```
It looks like that the API key is never used anywhere afterward, you do not need to store it in the configuration.

HofmeisterAn commented on src/Testcontainers.Qdrant/QdrantBuilder.cs:
```diff
@@ -0,0 +1,97 @@
+			}
+
+			return httpWaitStrategy;
+		});
```
Since we know the certificate, shouldn't we then check for the certificate? Or do we ignore it out of simplicity?

russcam replied:
Was ignoring it out of simplicity for the waiting strategy, but it could be checked

HofmeisterAn committed:
chore: Order csproj in sln

HofmeisterAn committed:
chore: Replace SolutionDir with relative path

HofmeisterAn committed:
chore: Remove BOM

HofmeisterAn committed:
chore: Fix minor repo inconsistencies

HofmeisterAn commented on tests/Testcontainers.Qdrant.Tests/X509CertificateGenerator.cs:
```diff
@@ -0,0 +1,52 @@
+{
+    public record PemCertificate(string Certificate, string PrivateKey);
+    
+    public static PemCertificate Generate(string subjectName)
```
Cannot we use .NET's built-in capabilities, like:

```csharp
var request = new CertificateRequest("CN=MyCertificate", RSA.Create(2048), HashAlgorithmName.SHA256, RSASignaturePadding.Pkcs1);
request.CertificateExtensions.Add(new X509BasicConstraintsExtension(false, false, 0, true));
var certificate = request.CreateSelfSigned(DateTimeOffset.UtcNow, DateTimeOffset.UtcNow.AddYears(1));
return new PemCertificate(certificate.ExportCertificatePem(), certificate.GetRSAPrivateKey().ExportPkcs8PrivateKeyPem());
```

russcam replied:
That's much simpler, will update 👍

HofmeisterAn reviewed (commented):
Thank you for the PR 🙏. I have reviewed it and made some minor changes to align with the repository standards. I have a some suggestions regarding the certificates. I am happy to merge the PR after we have sorted them out.

russcam committed:
Use CertificateRequest to generate PEM certificate

russcam committed:
Merge branch 'qdrant-container' of github.com:russcam/testcontainers-dotnet into qdrant-container

russcam committed:
remove BOM

eddumelendez commented:
It will be nice to add some docs for the module and how to use it.

russcam committed:
Update version and add docs

This commit updates the version of the image and the
official Qdrant client, and adds docs on how to use
the container.

russcam committed:
Merge branch 'develop' into qdrant-container

russcam committed:
Update to net9.0

russcam committed:
Add qdrant docs to configuration

russcam commented:
@HofmeisterAn @eddumelendez I've updated this PR with latest on develop and added documentation. Looking forward to seeing this merged.

russcam committed:
Configure qdrant in cicd.yml

HofmeisterAn commented:
I'll try to review it by the end of this week.

eddumelendez reviewed (commented):
LGTM

HofmeisterAn committed:
chore: Align repo standards

HofmeisterAn committed:
chore: Remove BOM

HofmeisterAn committed:
fix: Use correct id to include code snippet in docs

HofmeisterAn committed:
chore: Add note about PEM format

HofmeisterAn commented on src/Testcontainers.Qdrant/QdrantBuilder.cs:
```diff
@@ -0,0 +1,132 @@
+            .WithEnvironment("QDRANT__TLS__KEY", CertificateKeyFilePath)
+            .WithResourceMapping(Encoding.UTF8.GetBytes(certificate), CertificateFilePath)
+            .WithResourceMapping(Encoding.UTF8.GetBytes(certificateKey), CertificateKeyFilePath);
+    }
```
```suggestion
    /// <summary>
    /// Sets the API key to secure the instance.
    /// </summary>
    /// <param name="apiKey">The API key.</param>
    /// <returns>A configured instance of <see cref="QdrantBuilder" />.</returns>
    public QdrantBuilder WithApiKey(string apiKey)
    {
        return Merge(DockerResourceConfiguration, new QdrantConfiguration(apiKey: apiKey))
            .WithEnvironment("QDRANT__SERVICE__API_KEY", apiKey);
    }

    /// <summary>
    /// Sets the public certificate and private key to enable TLS.
    /// </summary>
    /// <param name="certificate">The public certificate in PEM format.</param>
    /// <param name="certificateKey">The private key associated with the certificate in PEM format.</param>
    /// <returns>A configured instance of <see cref="QdrantBuilder" />.</returns>
    public QdrantBuilder WithCertificate(string certificate, string certificateKey)
    {
        return Merge(DockerResourceConfiguration, new QdrantConfiguration(certificate: certificate, certificateKey: certificateKey))
            .WithEnvironment("QDRANT__SERVICE__ENABLE_TLS", "1")
            .WithEnvironment("QDRANT__TLS__CERT", CertificateFilePath)
            .WithEnvironment("QDRANT__TLS__KEY", CertificateKeyFilePath)
            .WithResourceMapping(Encoding.UTF8.GetBytes(certificate), CertificateFilePath)
            .WithResourceMapping(Encoding.UTF8.GetBytes(certificateKey), CertificateKeyFilePath);
    }
```

HofmeisterAn committed:
Update src/Testcontainers.Qdrant/QdrantBuilder.cs

HofmeisterAn commented on docs/modules/qdrant.md:
```diff
@@ -0,0 +1,75 @@
+
+    Please ensure that both the certificate and private key are provided in PEM format.
+
+## A Note To Developers
```
```suggestion
## Configure API key

To set and configure an API key, use the following container builder method:

=== "Configure the API key"
    ```csharp
    --8<-- "tests/Testcontainers.Qdrant.Tests/QdrantSecureContainerTest.cs:ConfigureQdrantContainerApiKey"
    ```

Make sure the underlying Qdrant HTTP or gRPC client adds the API key to the HTTP header or gRPC metadata:

=== "Configure the Qdrant client"
    ```csharp
    --8<-- "tests/Testcontainers.Qdrant.Tests/QdrantSecureContainerTest.cs:ConfigureQdrantClientApiKey"
    ```

## Configure TLS

The following example generates a self-signed certificate and configures the module to use TLS with the certificate and private key:

!!! note

    Please ensure that both the certificate and private key are provided in PEM format.

=== "Configure the TLS certificate"
    ```csharp
    --8<-- "tests/Testcontainers.Qdrant.Tests/QdrantSecureContainerTest.cs:ConfigureQdrantContainerCertificate"
    ```

The Qdrant client is configured to validate the TLS certificate using its thumbprint:

=== "Configure the Qdrant client"
    ```csharp
    --8<-- "tests/Testcontainers.Qdrant.Tests/QdrantSecureContainerTest.cs:ConfigureQdrantClientCertificate-1"

    --8<-- "tests/Testcontainers.Qdrant.Tests/QdrantSecureContainerTest.cs:ConfigureQdrantClientCertificate-2"
    ```

## A Note To Developers
```

HofmeisterAn committed:
Update docs/modules/qdrant.md

HofmeisterAn reviewed (approved):
Thanks ⭐.

HofmeisterAn committed:
Merge branch 'develop' into qdrant-container

russcam commented:
Thanks for merging @HofmeisterAn. Do you have a rough idea when this might be released?
